File: matrix-mult.cpp
#include <iostream>
#include <fstream>

using std::cout; using std::endl;

// change values according to matrices in input file
// columns_one must equal rows_two
const int NUMBER_OF_ROWS_ONE = 5;
const int NUMBER_OF_COLUMNS_ONE = 5;
const int NUMBER_OF_ROWS_TWO = 5;
const int NUMBER_OF_COLUMNS_TWO = 5;

int main(){
    std::ifstream matrixInput("matrix-input.txt");
    if(!matrixInput.is_open()){
        cout << "Error opening input file\n";
        return 1;
    }
    // create and fill matrix one
    // memory allocation done by programmer unlike javascript*
    int** matrixOne = new int*[NUMBER_OF_ROWS_ONE];
    for(int i = 0; i < NUMBER_OF_ROWS_ONE; ++i){
        matrixOne[i] = new int[NUMBER_OF_COLUMNS_ONE];
    }
    int num;
    for(int i = 0; i < NUMBER_OF_ROWS_ONE; ++i){
        for(int j = 0; j < NUMBER_OF_COLUMNS_ONE && matrixInput >> num; ++j){
            matrixOne[i][j] = num;
        }
    }
    // create matrix two
    int** matrixTwo = new int*[NUMBER_OF_ROWS_TWO];
    for(int i = 0; i < NUMBER_OF_ROWS_TWO; ++i){
        matrixTwo[i] = new int[NUMBER_OF_COLUMNS_TWO];
    }
    for(int i = 0; i < NUMBER_OF_ROWS_TWO; ++i){
        for(int j = 0; j < NUMBER_OF_COLUMNS_TWO && matrixInput >> num; ++j){
            matrixTwo[i][j] = num;
        }
    }
    // create matrix that gets product of one and two
    int** matrixProduct = new int*[NUMBER_OF_ROWS_ONE];
    for(int i = 0; i < NUMBER_OF_ROWS_ONE; ++i){
        matrixProduct[i] = new int[NUMBER_OF_COLUMNS_TWO];
    }
    // calculate product matrix
    
    int productElement = 0;
    for(int oneRow = 0; oneRow < NUMBER_OF_COLUMNS_ONE; ++oneRow){
        for(int twoColumn = 0; twoColumn < NUMBER_OF_COLUMNS_TWO; ++twoColumn){
            for(int oneColumn = 0, twoRow = 0; oneColumn < NUMBER_OF_COLUMNS_ONE
                && twoRow < NUMBER_OF_COLUMNS_TWO; ++oneColumn, ++twoRow){
                productElement += (matrixOne[oneRow][oneColumn] 
                    * matrixTwo[twoRow][twoColumn]);
            }
            matrixProduct[oneRow][twoColumn] = productElement;
            productElement = 0; 
        }
    }

    // output for debug purposes
    cout << "Matrix one:\n";
    for(int i = 0; i < NUMBER_OF_ROWS_ONE; ++i){
        for(int j = 0; j < NUMBER_OF_COLUMNS_ONE; ++j){
            cout << matrixOne[i][j] << "  ";
        }
        cout << endl;
    }
    cout << "Matrix two:\n";
    for(int i = 0; i < NUMBER_OF_ROWS_TWO; ++i){
        for(int j = 0; j < NUMBER_OF_COLUMNS_TWO; ++j){
            cout << matrixTwo[i][j] << "  ";
        }
        cout << endl;
    }
    cout << "Product matrix:\n";
    for(int i = 0; i < NUMBER_OF_ROWS_ONE; ++i){
        for(int j = 0; j < NUMBER_OF_COLUMNS_TWO; ++j){
            cout << matrixProduct[i][j] << "     ";
        }
        cout << endl;
    }

    
    
    matrixInput.close();
}

File: matrix-mult.js
// Creates matrices from input file and measures time it takes
// to find the product of the matrices
// Eric Stauss
// 4/28/2023

const NUMBER_OF_ROWS_ONE = 250;
const NUMBER_OF_COLUMNS_ONE = 250;
const NUMBER_OF_ROWS_TWO = 250;
const NUMBER_OF_COLUMNS_TWO = 250;

// dynamically typed unlike c++ so type of variable is associated
// with runtime values. Variable can be bound to different types
let matrixOne;
let matrixTwo;

// get numbers from text file to create matrices
fetch('matrix-input.txt')
    .then(response => response.text())
    .then(text => {
        const rows = text.trim().split('\n');
        matrixOne =
            rows.slice(0, NUMBER_OF_ROWS_ONE).map(row => row.trim().split(' ').map(Number));
        matrixTwo =
            rows.slice(NUMBER_OF_ROWS_ONE).map(row => row.trim().split(' ').map(Number));
        console.log(matrix1);
        console.log(matrix2);
    })
    .catch(error => console.log(error));

let matrixProduct = [];
for (let i = 0; i < NUMBER_OF_ROWS_ONE; ++i) {
    let row = [];
    for (let j = 0; j < NUMBER_OF_COLUMNS_TWO; ++j) {
        row.push(0);
    }
    matrixProduct.push(row);
}
console.log(matrixProduct);
window.onload = function () {
    const butt = document.getElementById('butt');
    butt.addEventListener('click', calcProduct);

    const directions = document.getElementById('directions');
    directions.textContent = `click start test to measure the time it takes to`
        + ` multiply a ${NUMBER_OF_ROWS_ONE}x${NUMBER_OF_COLUMNS_ONE} matrix with`
        + ` a ${NUMBER_OF_ROWS_TWO}x${NUMBER_OF_COLUMNS_TWO} matrix`

    const calcTime = document.getElementById('calc-time');

    // measures time it takes to calculate the product matrix
    // in basically exact same way as c++ file does except it
    // is significantly slower due to javascript being an 
    // intepretted language and c++ is compiled 
    function calcProduct() {
        const startTime = performance.now();
        let productElement = 0
        for (let oneRow = 0; oneRow < NUMBER_OF_COLUMNS_ONE; ++oneRow) {
            for (let twoColumn = 0; twoColumn < NUMBER_OF_COLUMNS_TWO; ++twoColumn) {
                for (let oneColumn = 0, twoRow = 0; oneColumn < NUMBER_OF_COLUMNS_ONE
                    && twoRow < NUMBER_OF_COLUMNS_TWO; ++oneColumn, ++twoRow) {
                    productElement += (matrixOne[oneRow][oneColumn]
                        * matrixTwo[twoRow][twoColumn]);
                }
                matrixProduct[oneRow][twoColumn] = productElement;
                productElement = 0;
            }
        }
        const endTime = performance.now();
        const duration = endTime - startTime;
        calcTime.textContent = `It took ${duration}ms to calculate the `
            + `${NUMBER_OF_ROWS_ONE}x${NUMBER_OF_COLUMNS_TWO} product matrix.`
    }
}
// notice how there is no calls to delete like in c++ file
// memory management is done automatically in javascript and not
// explicitly by programmer
